File: app/src/main/java/com/xjm/xxd/dribbird/utils/StatusBarCompat.java
package com.xjm.xxd.dribbird.utils;

import android.annotation.TargetApi;
import android.app.Activity;
import android.content.Context;
import android.content.res.Resources;
import android.os.Build;
import android.support.annotation.ColorRes;
import android.view.View;
import android.view.ViewGroup;

import com.xjm.xxd.dribbird.R;

/**
 * User : queda
 * Date : 17-3-19
 */

public class StatusBarCompat {

    private static final int COLOR_DEFAULT_ID = R.color.transparent;

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    public static void compat(Activity activity, @ColorRes int statusColor) {
        if (activity == null) {
            return;
        }
        Resources resources = activity.getResources();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            activity.getWindow().setStatusBarColor(resources.getColor(statusColor));
            return;
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT && Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            ViewGroup contentView = (ViewGroup) activity.findViewById(android.R.id.content);
            View statusBarView = new View(activity);
            ViewGroup.LayoutParams lp = new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                    getStatusBarHeight(activity));
            statusBarView.setBackgroundColor(resources.getColor(statusColor));
            contentView.addView(statusBarView, lp);
        }

    }

    public static void compat(Activity activity) {
        compat(activity, COLOR_DEFAULT_ID);
    }


    public static int getStatusBarHeight(Context context) {
        int result = 0;
        int resourceId = context.getResources().getIdentifier("status_bar_height", "dimen", "android");
        if (resourceId > 0) {
            result = context.getResources().getDimensionPixelSize(resourceId);
        }
        return result;
    }
}

File: app/src/main/java/com/xjm/xxd/dribbird/utils/StatusBarCompat.kt
package com.xjm.xxd.dribbird.utils

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.graphics.Color
import android.os.Build
import android.support.annotation.ColorRes
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import com.xjm.xxd.dribbird.R

/**
 * User : queda
 * Date : 17-3-19
 */

object StatusBarCompat {

    private const val COLOR_DEFAULT_ID = R.color.transparent

    @SuppressLint("ObsoleteSdkInt")
    @JvmOverloads
    fun compat(activity: Activity?, @ColorRes statusColor: Int = COLOR_DEFAULT_ID) {
        if (activity == null) {
            return
        }
        val resources = activity.resources
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            activity.window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
            activity.window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS)
            activity.window.statusBarColor = resources.getColor(statusColor, null)
            return
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT && Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            activity.window.addFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS)
            val contentView = activity.findViewById(android.R.id.content) as ViewGroup
            val statusBarView = View(activity)
            val lp = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                    getStatusBarHeight(activity))
            statusBarView.setBackgroundColor(resources.getColor(statusColor, null))
            contentView.addView(statusBarView, lp)
        }

    }

    private fun getStatusBarHeight(context: Context): Int {
        var result = 0
        val resourceId = context.resources.getIdentifier("status_bar_height", "dimen", "android")
        if (resourceId > 0) {
            result = context.resources.getDimensionPixelSize(resourceId)
        }
        return result
    }
}
